src/agent/advertising_proxy.cpp

```diff
@@ -0,0 +1,322 @@
+ *    ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *    LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *    CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *    SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *    INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *    CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *    ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *    POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   The file implements the Advertising Proxy.
+ */
+
+#include "agent/advertising_proxy.hpp"
+
+#include <string>
+
+#include <assert.h>
+
+#include "common/code_utils.hpp"
+#include "common/logging.hpp"
+
+#if OTBR_ENABLE_ADVERTISING_PROXY
+
+namespace otbr {
+
+// TODO: better to have SRP server provide separated service name labels.
+static otbrError SplitFullServiceName(const std::string &aFullName,
+                                      std::string &      aInstanceName,
+                                      std::string &      aType,
+                                      std::string &      aDomain)
+{
+    otbrError error = OTBR_ERROR_INVALID_ARGS;
+    size_t    dotPos[3];
+
+    dotPos[0] = aFullName.find_first_of('.');
+    VerifyOrExit(dotPos[0] != std::string::npos);
+
+    dotPos[1] = aFullName.find_first_of('.', dotPos[0] + 1);
+    VerifyOrExit(dotPos[1] != std::string::npos);
+
+    dotPos[2] = aFullName.find_first_of('.', dotPos[1] + 1);
+    VerifyOrExit(dotPos[2] != std::string::npos);
+
+    error         = OTBR_ERROR_NONE;
+    aInstanceName = aFullName.substr(0, dotPos[0]);
+    aType         = aFullName.substr(dotPos[0] + 1, dotPos[2] - dotPos[0] - 1);
+    aDomain       = aFullName.substr(dotPos[2] + 1, aFullName.size() - dotPos[2] - 1);
+
+exit:
+    return error;
+}
+
+// TODO: better to have the SRP server to provide separated host name.
+static otbrError SplitFullHostName(const std::string &aFullName, std::string &aHostName, std::string &aDomain)
+{
+    otbrError error = OTBR_ERROR_INVALID_ARGS;
+    size_t    dotPos;
+
+    dotPos = aFullName.find_first_of('.');
+    VerifyOrExit(dotPos != std::string::npos);
+
+    error     = OTBR_ERROR_NONE;
+    aHostName = aFullName.substr(0, dotPos);
+    aDomain   = aFullName.substr(dotPos + 1, aFullName.size() - dotPos - 1);
+
+exit:
+    return error;
+}
+
+AdvertisingProxy::AdvertisingProxy(Mdns::Publisher *aPublisher)
+    : mInstance(nullptr)
+    , mPublisher(aPublisher)
+{
+}
+
+otbrError AdvertisingProxy::Start(otInstance *aInstance)
+{
+    mInstance = aInstance;
+
+    otSrpServerSetServiceUpdateHandler(mInstance, AdvertisingHandler, this);
+
+    mPublisher->SetPublishServiceHandler(PublishServiceHandler, this);
+    mPublisher->SetPublishHostHandler(PublishHostHandler, this);
+
+    otbrLog(OTBR_LOG_INFO, "[adproxy] Started");
```

Comment:
```suggestion
    otbrLog(OTBR_LOG_INFO, "[SRP Advertising Proxy] Started");
```
nit: be more verbose ?

Reply:
I think `adproxy` is clear enough to distinguish Advertising Proxy logs from other modules. Besides, it is easier to grep `adproxy` that `SRP Advertising Proxy`.